pkg/proto/dispatch/v1/01_codec.go

```diff
@@ -20,38 +18,52 @@ import (
 const Name = "proto"
 
 type vtprotoMessage interface {
-	MarshalVT() ([]byte, error)
+	MarshalToSizedBufferVT(data []byte) (int, error)
 	UnmarshalVT([]byte) error
+	SizeVT() int
 }
 
-type vtprotoCodec struct{}
+type vtprotoCodec struct {
+	fallback encoding.CodecV2
+}
 
 func (vtprotoCodec) Name() string { return Name }
 
-func (vtprotoCodec) Marshal(v any) ([]byte, error) {
+func (c *vtprotoCodec) Marshal(v any) (mem.BufferSlice, error) {
 	if m, ok := v.(vtprotoMessage); ok {
-		return m.MarshalVT()
-	}
-
-	if m, ok := v.(proto.Message); ok {
-		return proto.Marshal(m)
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, size)
+			n, err := m.MarshalToSizedBufferVT(buf[:size])
```

Comment:
if we're allocating the buffer as size of `size`, why do we need to slice here and below?

Reply:
I cleaned this up a bit to try and make it more clear (and after checking guarantees from the pool implementations).
I'm being defensive for the scenario where the number of bytes written are not actually the exact size of the buffer, which looks possible based on the generated code.